### PR TITLE
.Net: Switch to using interfaces for search instead of query objects.

### PR DESCRIPTION
**Pull request by westey-m**
### Motivation and Context

As part of the vector search adr, we have agreed to switch to a different design for exposing different search experiences.

### Description

- Switch to individual interfaces for search.
- Making vector search part of the collection interface

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/Concepts/Memory/VectorSearch_Simple.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        // Create an embedding generation service.
+        var textEmbeddingGenerationService = new AzureOpenAITextEmbeddingGenerationService(
+                TestConfiguration.AzureOpenAIEmbeddings.DeploymentName,
+                TestConfiguration.AzureOpenAIEmbeddings.Endpoint,
+                TestConfiguration.AzureOpenAIEmbeddings.ApiKey);
+
+        // Initiate the docker container and construct the vector store.
+        await qdrantFixture.ManualInitializeAsync();
+        var vectorStore = new QdrantVectorStore(new QdrantClient("localhost"));
```
Comment (markwallace-microsoft) [inline dotnet/samples/Concepts/Memory/VectorSearch_Simple.cs:36]: Should the simplest example just use the volatile memory store?

You could consider adding two Facts and a utility method which takes a vector store instance as a parameter. Fact #1 would use volatile so zero friction and Fact #1 could use Quadrant.

File: dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs [hunk 2]
```diff
@@ -1,24 +1,27 @@
...
 
 using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 
 namespace Microsoft.SemanticKernel.Data;
 
 /// <summary>
-/// Interface for searching a vector store.
+/// Contains a method for doing a vector search on a vector store.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs:9]: performing a vector search on a vector store using a vector query

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs (hunk 2 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs]: Thanks, updated.

File: dotnet/samples/Concepts/Memory/VectorSearch_Simple.cs (hunk 1 above)
Comment (westey-m) [reply dotnet/samples/Concepts/Memory/VectorSearch_Simple.cs]: Yeah, makes sense.  For now I've just updated it to only use volatile, and I'll add more examples with the samples work item I have.

File: dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchVectorStoreRecordCollectionTests.cs [hunk 3]
```diff
@@ -608,16 +607,15 @@ public async Task CanSearchWithTextAndFilterAsync()
...
-            VectorSearchQuery.CreateQuery(
-                "search string",
-                new()
-                {
-                    Limit = 5,
-                    Offset = 3,
-                    Filter = filter,
-                    VectorFieldName = nameof(MultiPropsModel.Vector1)
-                }),
+        var searchResults = await sut.VectorizableTextSearchAsync(
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchVectorStoreRecordCollectionTests.cs:610]: A bit confused here why we have two names different, I prefer `Vectorized`.

`Vectorized` / `Vectorizable` ?

```suggestion
        var searchResults = await sut.VectorizableTextSearchAsync(
```

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs [hunk 4]
```diff
@@ -22,7 +22,7 @@ namespace Microsoft.SemanticKernel.Connectors.AzureAISearch;
 /// </summary>
 /// <typeparam name="TRecord">The data model to use for adding, updating and retrieving data from storage.</typeparam>
 #pragma warning disable CA1711 // Identifiers should not have incorrect suffix
-public sealed class AzureAISearchVectorStoreRecordCollection<TRecord> : IVectorStoreRecordCollection<string, TRecord>, IVectorSearch<TRecord>
+public sealed class AzureAISearchVectorStoreRecordCollection<TRecord> : IVectorStoreRecordCollection<string, TRecord>, IVectorizableTextSearch<TRecord>
```
Comment (SergeyMenshykh) [inline dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs:25]: Is it necessary to specify implementation of the `IVectorStoreRecordCollection` directly if it's already specified indirectly via `IVectorStoreRecordCollection` interface which inherits from the `IVectorStoreRecordCollection` interface?

File: dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs [hunk 5]
```diff
@@ -1,24 +1,27 @@
...
     /// Search the vector store for records that match the given embedding and filter.
     /// </summary>
-    /// <param name="vectorQuery">The vector to search with.</param>
+    /// <typeparam name="TVector">The type of the vector.</typeparam>
+    /// <param name="vector">The vector to search the store with.</param>
+    /// <param name="options">The options that control the behavior of the search.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
     /// <returns>The records found by the vector search, including their result scores.</returns>
-    IAsyncEnumerable<VectorSearchResult<TRecord>> SearchAsync(VectorSearchQuery vectorQuery, CancellationToken cancellationToken = default);
+    IAsyncEnumerable<VectorSearchResult<TRecord>> VectorizedSearchAsync<TVector>(
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs:23]: If I want to page over the results how will I do that?

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchVectorStoreRecordCollectionTests.cs (hunk 3 above)
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchVectorStoreRecordCollectionTests.cs:610]: Vectorized means that you are doing a search with a vector, as in the data is already vectorized.  Vectorizable means that you have something else that is vectorizable and will be vectorized lower down the stack, e.g. text.
Azure AI Search uses the same nomanclature, but we will probably revisit this before we go live again.

See e.g. the following for azure ai search.
https://learn.microsoft.com/en-us/javascript/api/@azure/search-documents/vectorizabletextquery?view=azure-node-latest

File: dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs (hunk 5 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/VectorSearch/IVectorizedSearch.cs:23]: You can use limit and offset together to do paging, e.g. for page two you can have this:
```csharp
await vectorSearch!.VectorizedSearchAsync(searchVector, new() { Limit = 3, Offset = 3, Filter = filter }).ToListAsync();
```

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs (hunk 4 above)
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs:25]: I suppose you mean is it necessary to specify `IVectorizedSearch` if it already inherits from `IVectorStoreRecordCollection`?  Note that there are two interfaces in play here.  `IVectorizedSearch` and `IVectorizableTextSearch`.  The first is a base for `IVectorStoreRecordCollection`, the second is not.  This was one of the main issues in the review, in that the options are either have the VectorizableText bit on the main interface, and throw were not supported, or have it separate.

**Pull request merged**